dashboard/src/api/auth-manager.ts

```diff
@@ -0,0 +1,138 @@
+import * as axios from 'axios';
+
+import { Api } from './index';
+import { AuthToken, CodeToken } from './types';
+
+class TokenStorage {
+    storeToken(token: string) {
+        if (!token) {
+            throw new Error('Cannot store empty token');
+        }
+        localStorage.setItem('AUTH_TOKEN', token);
+        localStorage.setItem('AUTH_TOKEN_STORED_AT', Date.now().toString());
+    }
+
+    getToken(): string | null {
+        return localStorage.getItem('AUTH_TOKEN');
+    }
+
+    clearToken() {
+        localStorage.removeItem('AUTH_TOKEN');
+        localStorage.removeItem('AUTH_TOKEN_STORED_AT');
+    }
+
+    getTokenAge(): number | null {
+        const timestampStr = localStorage.getItem('AUTH_TOKEN_STORED_AT');
+        if (!this.getToken() || !timestampStr) {
+            return null;
+        }
+        return Date.now() - Number(timestampStr);
+    }
+}
+
+const tokenStorage = new TokenStorage();
+
+export default class AuthManager {
+    /** Maximum age of the token in seconds before it is refreshed. */
+    public maxTokenAge: number = 5 * 60 * 1000;  // 5 minutes
+
+    private _api: Api;
+    private _axios: axios.AxiosInstance;
+    private _authInterceptorId: number | null = null;
+    private _tokenRefreshPromise: Promise<AuthToken> | null = null;
+
+    constructor(api: Api, axiosInstance: axios.AxiosInstance) {
+        this._api = api;
+        this._axios = axiosInstance;
+    }
+
+    checkExistingLogin(): Promise<AuthToken|null> {
+        return this.refreshLogin().then((authToken) => {
```

Comment:
Most if not all of the promises that I could find the in code has no return type (aka `any`)  when it comes to their return values. Might be something to consider adding.

Reply:
Do you mean the authToken being without a type definition in this line?

Reply:
Yes 🙂

Reply:
It's not actually `any`, since that's not even allowed (`"noImplicitAny": true,` in `tsconfig.json`). TypeScript just knows the type from the context.